compose/material3/material3/src/commonMain/kotlin/androidx/compose/material3/Text.kt

```diff
@@ -351,3 +357,62 @@ fun ProvideTextStyle(value: TextStyle, content: @Composable () -> Unit) {
     val mergedStyle = LocalTextStyle.current.merge(value)
     CompositionLocalProvider(LocalTextStyle provides mergedStyle, content = content)
 }
+
+/**
+ * Letter spacing doesn't play well with cursive writing systems, as it can break the connection
+ * between letters and make the text fragmented. A simple and performant solution to this problem is
+ * to find the first "letter" character in the text (using `isLetter()`) and set `letterSpacing` to
+ * 0 if that character belongs to a cursive writing system and `letterSpacing` is greater than 0.
+ */
+@Composable
+private fun rememberFixedLetterSpacing(
+    text: String,
+    letterSpacing: TextUnit,
+    styleLetterSpacing: TextUnit,
+) = remember(text, letterSpacing, styleLetterSpacing) {
+    when (letterSpacing.isSpecified) {
+        true -> {
+            when (letterSpacing.value == 0f) {
+                true -> letterSpacing
+                false -> if (text.isFirstLetterCursive()) 0f.sp else letterSpacing
+            }
+        }
+        false -> {
+            when (styleLetterSpacing.let { it.isUnspecified || it.value == 0f }) {
+                true -> TextUnit.Unspecified
+                false -> if (text.isFirstLetterCursive()) 0f.sp else styleLetterSpacing
+            }
+        }
+    }
+}
+
+/**
+ * Returns **`true`** if the first letter of the string belongs to a cursive writing system. A good
+ * example is the Persian language. Letter spacing doesn't play well with cursive writing systems,
+ * as it can break the connection between letters and make the text fragmented.
+ */
+@Stable
+private fun String.isFirstLetterCursive(): Boolean {
```

Comment:
In this version, we have defined a binarySearchRange extension for the range list that performs a binary search to check if a value is in any of the ranges. This drastically reduces the number of comparisons needed compared to the original approach of comparing against each range individually.

```
@Stable
private fun String.isFirstLetterCursive(): Boolean {
    val cursiveRanges = listOf(
        1536..1920,
        1984..2048,
        2144..2304,
        4096..4256,
        43488..43520,
        43616..43648,
        64336..65024,
        65136..65280
    )

    for (char in this) {
        if (!char.isLetter()) continue
        val c = char.code

        if (cursiveRanges.binarySearchRange(c)) {
            return true
        }
    }

    return false
}

private fun List<IntRange>.binarySearchRange(value: Int): Boolean {
    var low = 0
    var high = size - 1

    while (low <= high) {
        val mid = (low + high) / 2
        val range = this[mid]

        if (value in range) {
            return true
        }

        if (value < range.first) {
            high = mid - 1
        } else {
            low = mid + 1
        }
    }

    return false
}
```

Reply:
Hey @miguelalegria, thanks for the suggestion. The reason I didn't incorporate binary search and ranges is to avoid unnecessary allocations given the size of our search. I assume the current workaround would be more efficient. However, this pull request isn't probably going to be merged since the problem lies in lower layer of the system.